test/src/main/java/org/apache/accumulo/test/functional/FateInterleavingIT.java

```diff
@@ -0,0 +1,265 @@
+    try (AccumuloClient client = Accumulo.newClient().from(getClientProps()).build()) {
+      client.tableOperations().deleteRows(FATE_TRACKING_TABLE, null, null);
+    }
+  }
+
+  private void waitFor(ZooStore<Manager> store, long txid) throws Exception {
+    TStatus status;
+    do {
+      store.reserve(txid);
+      try {
+        status = store.getStatus(txid);
+      } finally {
+        store.unreserve(txid, Duration.ZERO);
+      }
+      if (status != TStatus.SUCCESSFUL) {
+        Thread.sleep(50);
+      }
+    } while (status != TStatus.SUCCESSFUL);
+  }
+
+  @Test
+  public void testDefaultInterleaving() throws Exception {
+
+    // Connect to the ZooKeeper that MAC is using and insert FATE operations
+    final String path = getCluster().getServerContext().getZooKeeperRoot() + Constants.ZFATE;
+    final ZooReaderWriter zrw = getCluster().getServerContext().getZooReaderWriter();
+    final ZooStore<Manager> store = new org.apache.accumulo.core.fate.ZooStore<>(path, zrw);
+
+    long txid = store.create();
+    store.reserve(txid);
+    try {
+      store.push(txid, new FirstOp());
+      store.setTransactionInfo(txid, TxInfo.TX_NAME, "TEST_ONE");
+      store.setStatus(txid, SUBMITTED);
+    } finally {
+      store.unreserve(txid, Duration.ZERO);
+    }
+
+    long txid2 = store.create();
+    store.reserve(txid2);
+    try {
+      store.push(txid2, new FirstOp());
+      store.setTransactionInfo(txid2, TxInfo.TX_NAME, "TEST_TWO");
+      store.setStatus(txid2, SUBMITTED);
+    } finally {
+      store.unreserve(txid2, Duration.ZERO);
+    }
+
+    long txid3 = store.create();
+    store.reserve(txid3);
+    try {
+      store.push(txid3, new FirstOp());
+      store.setTransactionInfo(txid3, TxInfo.TX_NAME, "TEST_THREE");
+      store.setStatus(txid3, SUBMITTED);
+    } finally {
+      store.unreserve(txid3, Duration.ZERO);
+    }
+
+    waitFor(store, txid);
+    waitFor(store, txid2);
+    waitFor(store, txid3);
+
+    // The execution order of the transactions is not according to their insertion
+    // order. However, we do know that the first step of each transaction will be
+    // executed before the second steps.
+    try (AccumuloClient client = Accumulo.newClient().from(getClientProps()).build()) {
+      Scanner scanner = client.createScanner(FATE_TRACKING_TABLE);
+      Iterator<Entry<Key,Value>> iter = scanner.iterator();
+
+      SortedMap<Key,Value> subset = new TreeMap<>();
+      AtomicInteger remaining = new AtomicInteger(6);
+
+      Stream.generate(() -> null)
+          .takeWhile(x -> (iter.hasNext() && remaining.getAndDecrement() > 0)).map(n -> iter.next())
+          .forEach(e -> subset.put(e.getKey(), e.getValue()));
+
+      assertTrue(
+          subset.values().stream().allMatch(v -> new String(v.get(), UTF_8).startsWith("FirstOp")));
+      assertEquals(2, subset.keySet().stream()
+          .filter(k -> k.getColumnFamily().toString().equals("TXID:" + txid)).count());
```

Comment:
For this and similar assertions, could also ensure that the isReady call occurs before the call call.